src/raftstore/store/fsm/peer.rs

```diff
@@ -1778,446 +1831,480 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+                info!("{} on split region at key {:?}.", self.peer.tag, split_keys);
+                self.on_prepare_split_region(region_epoch, split_keys, callback);
+            }
+            PeerMsg::RegionApproximateSize { size } => {
+                self.peer.approximate_size = Some(size);
+            }
+            PeerMsg::RegionApproximateKeys { keys } => {
+                self.peer.approximate_keys = Some(keys);
+            }
+            PeerMsg::CompactionDeclinedBytes(bytes) => {
+                self.peer.compaction_declined_bytes += bytes;
+                if self.peer.compaction_declined_bytes >= self.peer.cfg.region_split_check_diff.0 {
+                    UPDATE_REGION_SIZE_BY_COMPACTION_COUNTER.inc();
+                }
+            }
+            PeerMsg::GcSnap(keys) => self.on_gc_snap(keys),
+            PeerMsg::HalfSplitRegion {
+                region_epoch,
+                policy,
+            } => self.on_schedule_half_split_region(&region_epoch, policy),
+            PeerMsg::MergeResult { target, stale } => self.on_merge_result(target, stale),
+            PeerMsg::Start { state } => self.start(state),
+            PeerMsg::ClearStat => {
+                self.peer.approximate_size = None;
+                self.peer.approximate_keys = None;
+            }
+            PeerMsg::Noop => {}
         }
     }
 
-    fn on_ingest_sst_result(&mut self, ssts: Vec<SSTMeta>) {
-        for sst in &ssts {
-            let region_id = sst.get_region_id();
-            if let Some(region) = self.region_peers.get_mut(&region_id) {
-                region.size_diff_hint += sst.get_length();
+    pub fn poll(&mut self, receiver: &Receiver<PeerMsg>, buf: &mut Vec<PeerMsg>) -> Option<usize> {
+        let mut mark = None;
+        let mut shutdown = false;
+        if self.peer.pending_merge_apply.is_some() {
+            mark = Some(receiver.len());
+            if !self
+                .peer
+                .pending_merge_apply
+                .as_mut()
+                .unwrap()
+                .poller
+                .waken()
+            {
+                return mark;
+            }
+            if !self.resume_handling_pending_apply() {
+                return mark;
             }
+            mark = None;
         }
-
-        let task = CleanupSSTTask::DeleteSST { ssts };
-        if let Err(e) = self.cleanup_sst_worker.schedule(task) {
-            error!("schedule to delete ssts: {:?}", e);
+        while buf.len() < self.peer.cfg.messages_per_tick {
+            match receiver.try_recv() {
+                Ok(msg) => buf.push(msg),
+                Err(TryRecvError::Empty) => {
+                    mark = Some(0);
+                    break;
+                }
+                Err(TryRecvError::Disconnected) => {
+                    shutdown = true;
+                    mark = Some(0);
+                    break;
+                }
+            }
         }
+        for m in buf.drain(..) {
```

Comment:
If commit merge waiting, all the messages that after commit merge should also wait.